gpytorch/kernels/kernel.py

```diff
@@ -210,7 +225,78 @@ def __call__(self, x1, x2=None, **params):
         if x2_ is None:
             x2_ = x1_
 
-        return LazyEvaluatedKernelTensor(self, x1_, x2_, **params)
+        # Check batch_dims args
+        if settings.debug.on():
+            if batch_dims is not None:
+                if not isinstance(batch_dims, tuple) or (batch_dims != (0,) and batch_dims != (0, 2)):
+                    raise RuntimeError(
+                        "batch_dims currently accepts either None, (0,), or (0, 2). Got {}.".format(batch_dims)
+                    )
+
+        if diag:
+            res = super(Kernel, self).__call__(x1_, x2_, diag=True, batch_dims=batch_dims, **params)
+
+            # Did this Kernel eat the diag option?
+            # If it does not return a LazyEvaluatedKernelTensor, we can call diag on the output
+            if not isinstance(res, LazyEvaluatedKernelTensor):
+                if res.dim() == x1_.dim() and res.shape[-2:] == torch.Size((x1_.size(-2), x2_.size(-2))):
+                    res = res.diag()
+
+            # Now we'll make sure that the shape we're getting from diag makes sense
+            if settings.debug.on():
+                # If we used batch_dims...
+                shape = self.size(x1_, x2_)
+                if batch_dims == (0, 2):
+                    if len(shape) == 2:
+                        expected_shape = torch.Size((x1_.size(-1), shape[0]))
+                    else:
+                        expected_shape = torch.Size((shape[0] * x1_.size(-1), shape[1]))
+                    if res.shape != expected_shape:
+                        raise RuntimeError(
+                            "The kernel {} is not equipped to handle batch_dims=(0, 2) "
+                            "and diag. Expected size {}. Got size {}.".format(
+                                self.__class__.__name__, expected_shape, res.shape
+                            )
+                        )
+
+                # If we didn't use batch_dims...
+                else:
+                    expected_shape = shape[:-1]
+                    if res.shape != expected_shape:
+                        raise RuntimeError(
+                            "The kernel {} is not equipped to handle and diag. Expected size {}. "
+                            "Got size {}".format(self.__class__.__name__, expected_shape, res.shape)
+                        )
+            return res
+
+        else:
+            res = LazyEvaluatedKernelTensor(self, x1_, x2_, batch_dims=batch_dims, **params)
+
+            # Now we'll make sure that the shape we're getting makes sense
+            if settings.debug.on():
+                # If we used batch_dims...
+                shape = self.size(x1_, x2_)
+                if batch_dims == (0, 2):
```

Comment:
Even though we don't support multiple batch dimensions yet, it might be good to handle this correctly. Technically as coded, this breaks the contract for what batch_dims is supposed to mean. We can handle that in the issue that adds support for more than one batch dim, though.

Reply:
I don't understand what you mean.

Reply:
Forget I said anything. As long as we do the check against `(0,2)` I am happy for now.